### PR TITLE
Issue #SB-10785 fix: telemetry validation fixes 

### PR DESCRIPTION
**Pull request by raghuv45**

### CODE REVIEW

File: src/app/client/src/app/modules/core/services/data/data.service.ts [hunk 1]
```diff
@@ -48,6 +48,29 @@ export class DataService {
...
+   */
+  getWithHeaders(requestParam: RequestParam): Observable<ServerResponse> {
+    const httpOptions: HttpOptions = {
+      headers: requestParam.header ? requestParam.header : this.getHeader(),
+      params: requestParam.param,
+      observe: 'response'
+    };
+    return this.http.get(this.baseUrl + requestParam.url, httpOptions).pipe(
+      mergeMap((response: ServerResponseWithHeaders) => {
+        const data = response.body;
```
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/data/data.service.ts]: we dont need to declare var here

File: src/app/client/src/app/modules/core/services/device-register/device-register.service.ts [hunk 2]
```diff
@@ -1,19 +1,25 @@
...
 import { UUID } from 'angular2-uuid';
 import { HttpClient } from '@angular/common/http';
 import { DeviceDetectorService } from 'ngx-device-detector';
+import { Observable, timer, Subscription } from 'rxjs';
 
 @Injectable({
   providedIn: 'root'
 })
-export class DeviceRegisterService {
+export class DeviceRegisterService implements OnDestroy {
```
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/device-register/device-register.service.ts]: why do we have OnDestroy event in service. Dose angular call life cycle hook on services too.

File: src/app/client/src/app/modules/core/services/device-register/device-register.service.ts [hunk 3]
```diff
@@ -27,8 +33,22 @@ export class DeviceRegisterService {
     this.deviceRegisterApi = (<HTMLInputElement>document.getElementById('deviceRegisterApi'))
     && (<HTMLInputElement>document.getElementById('deviceRegisterApi')).value;
   }
+
+  onTimeOut () {
```
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/device-register/device-register.service.ts]: no need for this method

File: src/app/client/src/app/modules/core/services/device-register/device-register.service.ts [hunk 4]
```diff
@@ -27,8 +33,22 @@ export class DeviceRegisterService {
...
+    this.registerDevice(this.channel);
```
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/device-register/device-register.service.ts]: call this method in timer only

File: src/app/client/src/app/modules/core/services/device-register/device-register.service.ts [hunk 5]
```diff
@@ -27,8 +33,22 @@ export class DeviceRegisterService {
...
+  }
+
   registerDevice(channel: string, deviceId?: string) {
     console.log('calling registerDevice');
+    this.channel = channel;
+    // call register api every 24hrs
+    this.timer$ = timer(3.6e+6, 3.6e+6);
```
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/device-register/device-register.service.ts]: remove timer code from register device

File: src/app/client/src/app/modules/core/services/device-register/device-register.service.ts [hunk 6]
```diff
@@ -62,4 +82,8 @@ export class DeviceRegisterService {
       console.log('called device', err);
     });
   }
+
+  ngOnDestroy() {
```
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/device-register/device-register.service.ts]: dose service have ngOnDestroy event

File: src/app/client/src/app/modules/core/services/org-details/org-details.service.ts [hunk 7]
```diff
@@ -7,13 +7,15 @@ import { ContentService } from './../content/content.service';
...
 import { LearnerService } from './../learner/learner.service';
+
 @Injectable({
   providedIn: 'root'
 })
 export class OrgDetailsService {
 
   orgDetails: any;
   orgInfo: any;
+  _etsDates: any;
```
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/org-details/org-details.service.ts]: are we using this anywhere else other than telemetry init. If dont remove this

File: src/app/client/src/app/app.component.ts [hunk 8]
```diff
@@ -220,7 +220,8 @@ export class AppComponent implements OnInit {
           sid: this.userService.sessionId,
           channel: _.get(this.userProfile, 'rootOrg.hashTagId'),
           env: 'home',
-          enableValidation: environment.enableTelemetryValidation
+          enableValidation: environment.enableTelemetryValidation,
+          timeStampData: this.userService.getServerTime
```
Comment (anoophm) [inline src/app/client/src/app/app.component.ts]: rename to serverTime

File: src/app/client/src/app/modules/core/services/org-details/org-details.service.ts [hunk 9]
```diff
@@ -119,6 +122,19 @@ export class OrgDetailsService {
     return this.learnerService.get(systemSetting);
   }
 
+  /**
+  * serverresponse.ts is taken from header and not from api response ts, and format in IST
```
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/org-details/org-details.service.ts]: no need for function

File: src/app/client/src/app/modules/core/services/org-details/org-details.service.ts [hunk 10]
```diff
@@ -40,7 +42,8 @@ export class OrgDetailsService {
     if (this.orgDetails) {
       return observableOf(this.orgDetails);
     } else {
-      return this.publicDataService.post(option).pipe(mergeMap((data: ServerResponse) => {
+      return this.publicDataService.postWithHeaders(option).pipe(mergeMap((data: ServerResponse) => {
+        this.captureServerDate(data);
```
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/org-details/org-details.service.ts]: set the method here

File: src/app/client/src/app/modules/core/services/org-details/org-details.service.ts (hunk 7 above)
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/org-details/org-details.service.ts]: keep same name as telemetry lib

File: src/app/client/src/app/modules/core/services/user/user.service.ts [hunk 11]
```diff
@@ -147,8 +150,9 @@ export class UserService {
       url: `${this.config.urlConFig.URLS.USER.GET_PROFILE}${this.userid}`,
       param: this.config.urlConFig.params.userReadParam
     };
-    this.learnerService.get(option).subscribe(
+    this.learnerService.getWithHeaders(option).subscribe(
       (data: ServerResponse) => {
+        this.captureServerDate(data);
```
Comment (anoophm) [inline src/app/client/src/app/modules/core/services/user/user.service.ts]: no need for function

File: src/app/client/src/app/modules/core/services/data/data.service.ts (hunk 1 above)
Comment (raghuv45) [reply src/app/client/src/app/modules/core/services/data/data.service.ts]: done

File: src/app/client/src/app/modules/core/services/device-register/device-register.service.ts (hunk 2 above)
Comment (raghuv45) [reply src/app/client/src/app/modules/core/services/device-register/device-register.service.ts]: OnDestroy removed , seems to be not needed, lift cycle hooks works in services

File: src/app/client/src/app/modules/core/services/device-register/device-register.service.ts (hunk 4 above)
Comment (raghuv45) [reply src/app/client/src/app/modules/core/services/device-register/device-register.service.ts]: done

File: src/app/client/src/app/modules/core/services/device-register/device-register.service.ts (hunk 6 above)
Comment (raghuv45) [reply src/app/client/src/app/modules/core/services/device-register/device-register.service.ts]: yes

File: src/app/client/src/app/modules/core/services/org-details/org-details.service.ts (hunk 7 above)
Comment (raghuv45) [reply src/app/client/src/app/modules/core/services/org-details/org-details.service.ts]: done

File: src/app/client/src/app/modules/core/services/device-register/device-register.service.ts (hunk 3 above)
Comment (raghuv45) [reply src/app/client/src/app/modules/core/services/device-register/device-register.service.ts]: done

File: src/app/client/src/app/modules/core/services/user/user.service.ts (hunk 11 above)
Comment (raghuv45) [reply src/app/client/src/app/modules/core/services/user/user.service.ts]: this needs seperare function as it has some if checks and code comments better if it is seperate

File: src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts [hunk 12]
```diff
@@ -331,10 +333,12 @@ export class CoursePlayerComponent implements OnInit, OnDestroy {
     this.unsubscribe.complete();
   }
   private setTelemetryStartEndData() {
+    this.telemetryCdata = [{ 'type': 'course', 'id': this.courseId }, { 'type': 'batch', 'id': this.batchId }];
```
Comment (rajeevsathish) [inline src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts]: Type should be 'Course' and the 'batch' should be 'CourseBatch'

**Pull request merged**